R/PrestoQuery.R

```diff
@@ -205,13 +205,26 @@ PrestoQuery <- setRefClass("PrestoQuery",
       status <- 503L
       retries <- 3
       headers <- .request_headers(.conn)
+      extra_credentials <- .conn@extra.credentials
+      request_config <- .conn@request.config
+      statement <- enc2utf8(.statement)
       while (status == 503L || (retries > 0 && status >= 400L)) {
         wait()
-        post.response <- httr::POST(
-          url,
-          body = enc2utf8(.statement),
+        post_args <- list(
+          url = url,
+          body = statement,
           config = headers
         )
+
+        if (grepl("use.kerberos.headers=TRUE", extra_credentials)) {
+          post_args[["config"]] <- c(
+             post_args[["config"]],
+             request_config,
+             httr::authenticate(user = "", password = "", type = "gssnegotiate")
```

Comment:
So this is the extra header to pass to Presto server for Kerberos to work, right? I wonder why this cannot be passed directly by the user using the "request.config" arg of dbConnect()?

Reply:
Hello @jarodmeng . `request.config` is for `httr::config`, `httr::authenticate` is not a passable argument to `httr::config` (I tried and it returned an `Error in curl::handle_setopt(handle, .list = req$options) : Unknown option:` error).

Reply:
`http::authenticate()` calls `config()` (see [code link](https://github.com/r-lib/httr/blob/HEAD/R/authenticate.r)), so you should be able to use it directly rather than pass it to `httr::config()` again. i.e. Can you try `dbConnect(..., request.config = httr::authenticate(user = "", password = "", type = "gssnegotiate"), ... )`?

Reply:
Hello @jarodmeng . It won't work as it will override the httr::config setup required for Kerberos connection:

```
> presto_kerberos_ica <- RPresto::dbConnect(
+     drv = RPresto::Presto(),
+     use.trino.headers = TRUE,
+     catalog = "silver",
+     schema = "default",
+     user = USER,
+     host = KERBEROS_HOST,
+     port = KERBEROS_PORT,
+     extra.credentials = "use.kerberos.headers=TRUE",
+     request.config = httr::config(service_name = "presto",ssl_verifypeer=FALSE,ssl_verifyhost=FALSE)
+ )
> presto_kerberos_ica
<PrestoConnection: KERBEROS_HOST:KERBEROS_PORT>
Catalog: silver
Schema: default
User: USER
Source: RPresto
Session Time Zone: UTC
Output Time Zone: UTC
Extra Credentials: use.kerberos.headers=TRUE
BIGINT cast to: integer
> presto_kerberos_ica <- RPresto::dbConnect(
+     drv = RPresto::Presto(),
+     use.trino.headers = TRUE,
+     catalog = "silver",
+     schema = "default",
+     user = USER,
+     host = KERBEROS_HOST,
+     port = KERBEROS_PORT,
+     extra.credentials = "use.kerberos.headers=TRUE",
+     request.config = httr::authenticate(user = "", password = "", type = "gssnegotiate")
+ )
Error in .check.response.status(.response) :
Received error response (HTTP 401): Unauthorized
```

Reply:
I see. How about you use the `request_combine()` function (unexported) from `httr` to combine the `httr::authenticate()` and `httr::config()` before passing to the `request.config` arg?

i.e.

```
RPresto::dbConnect(
  drv = RPresto::Presto(),
  use.trino.headers = TRUE,
  catalog = "silver",
  schema = "default",
  user = USER,
  host = KERBEROS_HOST,
  port = KERBEROS_PORT,
  request.config = httr:::request_combine(
    httr::authenticate(user = "", password = "", type = "gssnegotiate"),
    httr::config(service_name = "presto", ssl_verifypeer=FALSE, ssl_verifyhost=FALSE)
  )
)
```

If this approach works, we can discuss what's the best way to incorporate this functionality into RPresto.

Reply:
Great, it works! Thanks @jarodmeng . With this, the `httr::authenticate` argument can also come from RPresto user/client.
```
> presto_kerberos_ica <- RPresto::dbConnect(
+     drv = RPresto::Presto(),
+     use.trino.headers = TRUE,
+     catalog = "silver",
+     schema = "default",
+     user = USER,
+     host = KERBEROS_HOST,
+     port = KERBEROS_PORT,
+     request.config = httr:::request_combine(
+         httr::authenticate(user = "", password = "", type = "gssnegotiate"),
+         httr::config(service_name = "presto", ssl_verifypeer=FALSE, ssl_verifyhost=FALSE)
+     )
+ )
> presto_kerberos_ica
<PrestoConnection: KERBEROS_HOST:KERBEROS_PORT>
Catalog: silver
Schema: default
User: USER
Source: RPresto
Session Time Zone: UTC
Output Time Zone: UTC
Extra Credentials: 
BIGINT cast to: integer
```

Reply:
Glad it works. Actually it seems that `httr::authenticate(user = "", password = "", type = "gssnegotiate")` basically translates into `httr::config(httpauth = 4)`. Can you try the following `request.config` value to see if the simplified version still works.

```
RPresto::dbConnect(
  drv = RPresto::Presto(),
  use.trino.headers = TRUE,
  catalog = "silver",
  schema = "default",
  user = USER,
  host = KERBEROS_HOST,
  port = KERBEROS_PORT,
  request.config = httr::config(httpauth = 4, service_name = "presto", ssl_verifypeer=FALSE, ssl_verifyhost=FALSE)
)
```

Reply:
Hello @jarodmeng . That option doesn't work.
```
> presto_kerberos_ica <- RPresto::dbConnect(
+     drv = RPresto::Presto(),
+     use.trino.headers = TRUE,
+     catalog = "silver",
+     schema = "default",
+     user = USER,
+     host = KERBEROS_HOST,
+     port = KERBEROS_PORT,
+     request.config = httr::config(httpauth = 4, service_name = "presto", ssl_verifypeer=FALSE, ssl_verifyhost=FALSE)
+ )
Error in .check.response.status(.response) :
Received error response (HTTP 401): Unauthorized
```

Reply:
It seems that the `userpwd` field of the config is critical too.

Try this block

```
RPresto::dbConnect(
  drv = RPresto::Presto(),
  use.trino.headers = TRUE,
  catalog = "silver",
  schema = "default",
  user = USER,
  host = KERBEROS_HOST,
  port = KERBEROS_PORT,
  request.config = httr::config(
    httpauth = 4,
    userpwd = ":",
    service_name = "presto",
    ssl_verifypeer = FALSE,
    ssl_verifyhost=FALSE
  )
)
```

Appendix: `httr::config()` reprex

``` r
httr:::request_combine(
  httr::authenticate(user = "", password = "", type = "gssnegotiate"),
  httr::config(
    service_name = "presto",
    ssl_verifypeer = FALSE,
    ssl_verifyhost = FALSE
  )
)
#> <request>
#> Options:
#> * httpauth: 4
#> * userpwd: :
#> * service_name: presto
#> * ssl_verifypeer: FALSE
#> * ssl_verifyhost: FALSE

httr::config(
  httpauth = 4,
  userpwd = ":",
  service_name = "presto",
  ssl_verifypeer = FALSE,
  ssl_verifyhost = FALSE
)
#> <request>
#> Options:
#> * httpauth: 4
#> * userpwd: :
#> * service_name: presto
#> * ssl_verifypeer: FALSE
#> * ssl_verifyhost: FALSE
```

<sup>Created on 2022-11-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Hello @jarodmeng . Great! This option works! We can now simplify this PR to just pass `request.config` argument.
```
> presto_kerberos_ica <- RPresto::dbConnect(
+     drv = RPresto::Presto(),
+     use.trino.headers = TRUE,
+     catalog = "silver",
+     schema = "default",
+     user = USER,
+     host = KERBEROS_HOST,
+     port = KERBEROS_PORT,
+     request.config = httr::config(
+         httpauth = 4,
+         userpwd = ":",
+         service_name = "presto",
+         ssl_verifypeer = FALSE,
+         ssl_verifyhost=FALSE
+     )
+ )
> presto_kerberos_ica
<PrestoConnection: KERBEROS_HOST:KERBEROS_PORT>
Catalog: silver
Schema: default
User: USER
Source: RPresto
Session Time Zone: UTC
Output Time Zone: UTC
Extra Credentials: 
BIGINT cast to: integer
```

Reply:
Awesome. I will add a function to wrap the `httr::config(...)` so that Kerberos users can call the function and supply the output to `request.config`. You can close this PR now. Thanks for the suggestion! It's really helpful, especially since I haven't got Kerberos to work on my development machine yet.

Reply:
Thanks a lot @jarodmeng for the review and the help!